pennylane/interfaces/tensorflow.py

```diff
@@ -14,85 +14,134 @@
+array([[0.2, 0. ],
+       [0. , 0.4]], dtype=float32)>
+
+As you can see, we got 3 calls to ``vjp`` instead of 2, and the first calls have identical ``dy``. We do not want
+to have to perform this extra call.
+
 """
-# pylint: disable=too-many-arguments,too-many-branches
-import inspect
+# pylint: disable=unused-argument
 import logging
+import warnings
 
 import tensorflow as tf
 from tensorflow.python.eager import context
 
 import pennylane as qml
 from pennylane.measurements import Shots
-from pennylane.transforms import convert_to_numpy_parameters
 
 logger = logging.getLogger(__name__)
 logger.addHandler(logging.NullHandler())
 
 
-def _set_copy_and_unwrap_tape(t, a, unwrap=True):
-    """Copy a given tape with operations and set parameters"""
-    tc = t.bind_new_parameters(a, list(range(len(a))))
-    return convert_to_numpy_parameters(tc) if unwrap else tc
-
-
-def set_parameters_on_copy_and_unwrap(tapes, params, unwrap=True):
+def set_parameters_on_copy(tapes, params):
     """Copy a set of tapes with operations and set parameters"""
-    return tuple(_set_copy_and_unwrap_tape(t, a, unwrap=unwrap) for t, a in zip(tapes, params))
-
-
-def _compute_vjp(dy, jacs, multi_measurements, has_partitioned_shots):
-    # compute the vector-Jacobian product dy @ jac
-    # for a list of dy's and Jacobian matrices.
-
-    if logger.isEnabledFor(logging.DEBUG):
-        logger.debug(
-            "Entry with args=(dy=%s, jacs=%s, multi_measurements=%s, shots=%s) called by=%s",
-            dy,
-            jacs,
-            multi_measurements,
-            has_partitioned_shots,
-            "::L".join(str(i) for i in inspect.getouterframes(inspect.currentframe(), 2)[1][1:3]),
-        )
-
-    vjps = []
-
-    for dy_, jac_, multi in zip(dy, jacs, multi_measurements):
-        dy_ = dy_ if has_partitioned_shots else (dy_,)
-        jac_ = jac_ if has_partitioned_shots else (jac_,)
-
-        shot_vjps = []
-        for d, j in zip(dy_, jac_):
-            if multi:
-                shot_vjps.append(qml.gradients.compute_vjp_multi(d, j))
-            else:
-                shot_vjps.append(qml.gradients.compute_vjp_single(d, j))
-
-        vjp = qml.math.sum(qml.math.stack(shot_vjps), 0)
-
-        if not context.executing_eagerly():
-            vjp = qml.math.unstack(vjp)
-
-        vjps.extend(vjp)
-
-    return vjps
+    return tuple(t.bind_new_parameters(a, list(range(len(a)))) for t, a in zip(tapes, params))
 
 
 def _to_tensors(x):
     """
     Convert a nested tuple structure of arrays into a nested tuple
     structure of TF tensors
     """
-    if isinstance(x, dict) or isinstance(x, list) and all(isinstance(i, dict) for i in x):
+    if x is None or isinstance(x, dict):
         # qml.counts returns a dict (list of dicts when broadcasted), can't form a valid tensor
         return x
 
-    if isinstance(x, tuple):
+    if isinstance(x, (tuple, list)):
         return tuple(_to_tensors(x_) for x_ in x)

```

Comment:
no risk converting lists to tuples here?

Reply:
We are pretty haphazard about whether results are lists or tuples in a lot of places.  I personally prefer using a `tuple`, so out of the two, I'd prefer tuples. 

But we could also just use `type(x)` instead of `tuple`.

Reply:
doesn't seem to break anything so I'd say it's fine.